File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/block/NativeBlockProvider.kt
package io.nativeblocks.runtime.api.provider.block

import androidx.compose.runtime.Composable

/**
 * Provides and manages native blocks within the framework.
 */
internal class NativeBlockProvider {

    /**
     * Map of block types to their corresponding implementations.
     */
    private val blocks = mutableMapOf<String, BlockComposable>()

    /**
     * Fallback block implementation to be used when a specific block type is not found.
     */
    private var fallbackBlock: (@Composable (keyType: String, key: String) -> Unit)? = null

    /**
     * Registers a new block implementation for a specific block type.
     * @param blockType The type of the block.
     * @param block The block implementation to register.
     */
    fun provideBlock(blockType: String, block: BlockComposable) {
        blocks[blockType] = block
    }

    /**
     * Retrieves all registered block implementations.
     * @return A map of block types to their implementations.
     */
    fun getProvidedBlocks(): Map<String, BlockComposable> {
        return blocks.toMap()
    }

    /**
     * Registers a fallback block implementation to be used when a specific block type is not found.
     * @param block The fallback block implementation to register.
     */
    fun onFallbackBlock(block: @Composable (keyType: String, key: String) -> Unit) {
        fallbackBlock = block
    }

    /**
     * Retrieves the fallback block implementation.
     * @return The fallback block implementation, or null if none is registered.
     */
    fun getFallbackBlock(): @Composable ((keyType: String, key: String) -> Unit)? {
        return fallbackBlock
    }

}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/block/NativeBlockProvider.swift
import Foundation
import SwiftUI

/// A singleton class responsible for managing and providing blocks within the Nativeblocks SDK.
/// The `NativeBlockProvider` class maintains a registry of blocks that can be dynamically provided based on the block type.
public class NativeBlockProvider {
    /// A dictionary to store the blocks, where the key is the block type and the value is the corresponding block implementation.
    private var blocks = [String: ((BlockContext) -> any View)]()

    /// Fallback block implementation to be used when a specific block type is not found
    private var fallbackBlock: ((String, String) -> any View)? = nil

    /// Registers a new block with the specified block key type.
    /// - Parameters:
    ///   - blockType: The key type for the block to be registered.
    ///   - block: The block instance that implements the `INativeBlock` protocol.
    func provideBlock(blockType: String, block: @escaping ((BlockContext) -> any View)) {
        blocks[blockType] = block
    }

    /// Sets the fallback block to be used when no matching block is registered or available.
    /// This block will be displayed in place of unsupported or unrecognized blocks.
    ///
    /// - Parameter block: An instance conforming to the `any INativeFallbackBlock` protocol to be used as the fallback.
    func onFallbackBlock(block: @escaping (String, String) -> any View) {
        fallbackBlock = block
    }

    /// Retrieves the dictionary of all registered blocks.
    /// - Returns: A dictionary containing all registered blocks where the key is the block type and the value is the block instance.
    func getProvidedBlocks() -> [String: ((BlockContext) -> any View)] {
        return blocks
    }

    /// Retrieves the fallback block.
    /// - Returns: Registered block
    func getFallbackBlock() -> ((String, String) -> any View)? {
        return fallbackBlock
    }
}
